src/mem3_shards.erl

```diff
@@ -242,7 +242,48 @@ terminate(_Reason, #st{changes_pid=Pid}) ->
     exit(Pid, kill),
     ok.
 
-code_change(_OldVsn, #st{}=St, _Extra) ->
+code_change(OldVsn, {st, MaxSize, CurSize, ChangesPid} = St, _Extra) ->
+    twig:log(notice, "~p code_change ~p ~p", [?MODULE, OldVsn, St]),
+    upgrade(MaxSize, CurSize, ChangesPid);
+
+code_change(OldVsn, {st, MaxSize, CurSize, ChangesPid, _, _} = St, _Extra) ->
```

Comment:
given this matches the arity of the current state record, this clause will be executed on all future hot upgrades, activating the downgrade code and crashing. You'll need to match the OldVsn here as `{down, 2}` to be sure we only do this on actual downgrade.